wxSVGFileDC improvements.

MaartenBent opened the pull request:
I used the screens in the drawing sample and tried to fix most of the issues when saving them as SVG, including:
- Enabled usage of clipping regions.
- Correctly draw polypolygons.
- Draw lines as one long line instead of many short lines.
- Drawing text improvements (position, multi-line, underlined, strike-through).
- Support more brush and pen styles.
- Add Saving as SVG to drawing sample.
- Implemented Clear().
- Set the SVG title.
- Produce valid svg/xml.
- Correctly draw ellipses and arcs.

The commit messages have some more info.

There are a lot of commits to make the individual changes more obvious. I don't have any objection to squashing them together.

Screenshots as wxDC, SVG (old), SVG (new). I converted the SVGs to PNGs, hence the lower quality.

Brushes:
<a href="https://cloud.githubusercontent.com/assets/8088070/13541307/c2726022-e25b-11e5-85d2-e66c85ee9d37.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541307/c2726022-e25b-11e5-85d2-e66c85ee9d37.png" width="150" alt="Brushes wxDC"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/13541319/d419afb0-e25b-11e5-939c-adf973f1ab3a.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541319/d419afb0-e25b-11e5-939c-adf973f1ab3a.png" width="150" alt="Brushes SVG Before"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/13541316/d1cd482a-e25b-11e5-8ea9-b51c160f65b8.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541316/d1cd482a-e25b-11e5-8ea9-b51c160f65b8.png" width="150" alt="Brushes SVG After"/></a>

Lines:
<a href="https://cloud.githubusercontent.com/assets/8088070/13541320/d6b33caa-e25b-11e5-800e-f62b59125ae3.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541320/d6b33caa-e25b-11e5-800e-f62b59125ae3.png" width="150" alt="Lines wxDC"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/13541325/dbc10ec0-e25b-11e5-93ca-81f06d9804ba.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541325/dbc10ec0-e25b-11e5-93ca-81f06d9804ba.png" width="150" alt="Lines SVG Before"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/13541321/d9416438-e25b-11e5-8870-52e59c64a4cd.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541321/d9416438-e25b-11e5-8870-52e59c64a4cd.png" width="150" alt="Lines SVG After"/></a>

Polygon:
<a href="https://cloud.githubusercontent.com/assets/8088070/13541352/07d1f560-e25c-11e5-881c-2c6cb0f6dc72.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541352/07d1f560-e25c-11e5-881c-2c6cb0f6dc72.png" width="150" alt="Polygon wxDC"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/13541358/0bf336fe-e25c-11e5-9713-0b8dd8c647be.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541358/0bf336fe-e25c-11e5-9713-0b8dd8c647be.png" width="150" alt="Polygon SVG Before"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/13541356/09ffcd26-e25c-11e5-8924-33b972d5f71d.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541356/09ffcd26-e25c-11e5-8924-33b972d5f71d.png" width="150" alt="Polygon SVG After"/></a>

Regions:
<a href="https://cloud.githubusercontent.com/assets/8088070/13541361/0ecd851e-e25c-11e5-842c-ff93b4b37f56.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541361/0ecd851e-e25c-11e5-842c-ff93b4b37f56.png" width="150" alt="Regions wxDC"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/13541367/122f1cc2-e25c-11e5-8e44-e6b5f9373421.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541367/122f1cc2-e25c-11e5-8e44-e6b5f9373421.png" width="150" alt="Regions SVG Before"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/13541364/1108645c-e25c-11e5-94d5-b1d98254a7c3.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541364/1108645c-e25c-11e5-94d5-b1d98254a7c3.png" width="150" alt="Regions SVG After"/></a>

Text:
<a href="https://cloud.githubusercontent.com/assets/8088070/13541371/1552a608-e25c-11e5-8abf-90e2e31daa1e.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541371/1552a608-e25c-11e5-8abf-90e2e31daa1e.png" width="150" alt="Text wxDC"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/13541379/1ce514be-e25c-11e5-92da-ce2419b8bb1e.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541379/1ce514be-e25c-11e5-92da-ce2419b8bb1e.png" width="150" alt="Text SVG Before"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/13541374/170331e8-e25c-11e5-8058-3713314c25e3.png"><img src="https://cloud.githubusercontent.com/assets/8088070/13541374/170331e8-e25c-11e5-8058-3713314c25e3.png" width="150" alt="Text SVG After"/></a>

Circles:
<a href="https://cloud.githubusercontent.com/assets/8088070/15799642/600c95ec-2a63-11e6-91f0-017787c66f73.png"><img src="https://cloud.githubusercontent.com/assets/8088070/15799642/600c95ec-2a63-11e6-91f0-017787c66f73.png" width="150" alt="Circles wxDC"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/15799643/649cfe3a-2a63-11e6-85ca-4ae72f800bf2.png"><img src="https://cloud.githubusercontent.com/assets/8088070/15799643/649cfe3a-2a63-11e6-85ca-4ae72f800bf2.png" width="150" alt="Circles SVG Before"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/15799644/67413da4-2a63-11e6-89e6-d5f9bcf72944.png"><img src="https://cloud.githubusercontent.com/assets/8088070/15799644/67413da4-2a63-11e6-89e6-d5f9bcf72944.png" width="150" alt="Circles SVG After"/></a>

SVG Arc Example:
<a href="https://cloud.githubusercontent.com/assets/8088070/15799650/a768ec9c-2a63-11e6-982b-c78de6136f58.png"><img src="https://cloud.githubusercontent.com/assets/8088070/15799650/a768ec9c-2a63-11e6-982b-c78de6136f58.png" width="150" alt="Arcs wxDC"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/15799651/aa102ee2-2a63-11e6-9a24-f8fd3adbc294.png"><img src="https://cloud.githubusercontent.com/assets/8088070/15799651/aa102ee2-2a63-11e6-9a24-f8fd3adbc294.png" width="150" alt="Arcs SVG Before"/></a><a href="https://cloud.githubusercontent.com/assets/8088070/15799652/ac7da5ce-2a63-11e6-822d-f9c49cacf5b6.png"><img src="https://cloud.githubusercontent.com/assets/8088070/15799652/ac7da5ce-2a63-11e6-822d-f9c49cacf5b6.png" width="150" alt="Arcs SVG After"/></a>

MaartenBent commented:
I had to remove the ```title``` attribute for images. It is possible to add a title using ```<title>text</title>``` inside the ```image``` tag, but I don't see any benefits of adding the current (fixed) title.

crnd commented on include/wx/dcsvg.h:
```diff
@@ -66,7 +66,8 @@ class WXDLLIMPEXP_CORE wxSVGFileDCImpl : public wxDCImpl
 public:
     wxSVGFileDCImpl( wxSVGFileDC *owner, const wxString &filename,
-                     int width=320, int height=240, double dpi=72.0 );
+                     int width=320, int height=240, double dpi=72.0,
```
Should try to keep the code style consistent by using spaces around the assignment operator:
```c++
int width = 320, int height = 240, double dpi = 72.0

MaartenBent commented:
Some modifications:
- Use spaces around assignment operator
- Use `switch` instead of `if else if else if else` for pen/brush styles
- Split multi-line text and font changes into separate commits.

vadz commented on include/wx/dcsvg.h:
```diff
@@ -66,7 +66,8 @@ class WXDLLIMPEXP_CORE wxSVGFileDCImpl : public wxDCImpl
     wxSVGFileDCImpl( wxSVGFileDC *owner, const wxString &filename,
-                     int width=320, int height=240, double dpi=72.0 );
+                     int width = 320, int height = 240, double dpi = 72.0,
+                     const wxString &title = wxT("") );
```
Hardly very important but it's better to use `= wxString()` for the default value (in spite of the fact that most of the existing code doesn't do it).

vadz commented on interface/wx/dcsvg.h:
```diff
@@ -40,9 +40,10 @@ class wxSVGFileDC : public wxDC
     /**
         Initializes a wxSVGFileDC with the given @a f filename with the given
-        @a Width and @a Height at @a dpi resolution.
+        @a Width and @a Height at @a dpi resolution, and an optional @a title.
```
Would be nice to explain how is the title used, at least to me (who doesn't know much about SVG) this is not really clear.

vadz commented on samples/drawing/drawing.cpp:
```diff
@@ -2130,14 +2136,40 @@ void MyFrame::OnSave(wxCommandEvent& WXUNUSED(event))
+            m_canvas->UseGraphicRenderer(NULL);
+#endif
+            wxSVGFileDC svgdc(dlg.GetPath(), width, height);
+            wxSVGBitmapHandler* bitmapHandler = new wxSVGBitmapEmbedHandler();
```
Do we need to create this on the heap? Also, it looks like `svgdc` is left with a dangling pointer currently (which it hopefully doesn't use, but still...). I'd create the bitmap handler before `svgdc` on the stack, so it's automatically destroyed after it.

vadz commented on src/common/dcsvg.cpp:
```diff
@@ -108,6 +118,151 @@ wxString wxBrushString(wxColour c, int style = wxBRUSHSTYLE_SOLID)
+            s += wxT("\" ");
+            break;
+        }
+        default:
```
This doesn't seem to be necessary.

vadz replied:
Just to make this more clear: it's a good idea to write exhaustive switch statements over enums, i.e. take into account all possible enum values. So in this case I'd rather all of `case wxPENSTYLE_XXX_HATCH:` explicitly here, even if we don't handle them and also add `case wxPENSTYLE_INVALID:` even if it can't really occur here (or can it? I'm not even sure, the function doesn't document its preconditions) and remove `default:`.

The advantage is that if some new `wxPENSTYLE_ZZZ` is added later, the compiler will remind us to add support for it here too.

vadz commented on src/common/dcsvg.cpp:
```diff
@@ -108,6 +118,151 @@ wxString wxBrushString(wxColour c, int style = wxBRUSHSTYLE_SOLID)
+    {
+        case wxBRUSHSTYLE_FDIAGONAL_HATCH:
+        {
+            wxString patternName = wxT("FdiagonalHatch") + brush.GetColour().GetAsString(wxC2S_HTML_SYNTAX).substr(1);
```
It's a bit painful to repeat all this code, couldn't we just initialize `patternName` in the switch and do everything else outside of it? We seem to be doing the same thing for all cases or am I missing something?

vadz commented:
Thanks, this definitely looks useful and it's great to have better SVG output support. If we could avoid adding more duplicated code (and I know there is already plenty of it), it would be really perfect...

MaartenBent commented:
I addressed your comments in the new commits.

The title is metadata, just like `html` and `pdf` can have a title. If you open a `svg` file in a browser, the title is shown as tab-name.

`wxSVGFileDC` takes ownership of the bitmap handler according to the documentation, so it should also take care of deleting it. I changed the usage of pointers to smart pointers to simplify this.

`default:` is needed to prevent compilers from complaining about  missing enumeration values:
```warning: enumeration value 'xxx' not handled in switch [-Wswitch]```

vadz commented on src/common/dcsvg.cpp:
```diff
@@ -108,6 +118,127 @@ wxString wxBrushString(wxColour c, int style = wxBRUSHSTYLE_SOLID)
+    float w = pen.GetWidth();
+    if (pen.GetWidth() == 0)
+        w = 1;
+    w = w / 3;
```
I wonder where do all these factors (3 here, 2, 5, 8, 10 below) come from? It would be nice to have a comment explaining it...

vadz commented on src/common/dcsvg.cpp:
```diff
@@ -313,10 +458,27 @@ void wxSVGFileDCImpl::DoDrawLine (wxCoord x1, wxCoord y1, wxCoord x2, wxCoord y2
-    {
-        DoDrawLine ( points [i-1].x + xoffset, points [i-1].y + yoffset,
-            points [ i ].x + xoffset, points [ i ].y + yoffset );
+    if (n > 1) {
```
Style: `{` should be on its own line.

vadz commented on src/common/dcsvg.cpp:
```diff
@@ -490,6 +676,48 @@ void wxSVGFileDCImpl::DoDrawPolygon(int n, const wxPoint points[],
+    }
+
+    wxPoint* pts;
+    pts = new wxPoint[totalPts + n];
```
Could we please use `wxScopedArray<wxPoint> pts(totalPts + n)` instead for the peace of mind?

vadz commented on src/common/dcsvg.cpp:
```diff
@@ -743,7 +973,7 @@ void wxSVGFileDCImpl::NewGraphicsIfNeeded()
     m_graphics_changed = false;
 
-    write(wxS("</g>\n"));
+    write(wxT("</g>\n"));
```
Why? We do have an awful mess of `wxT()` and `wxS()` here and I don't say we should necessarily clean it up immediately, but at least let's not change correct latter ones into less correct former ones. And `wxS()` is better because `wxString` can be constructed from it more efficiently.

vadz commented on include/wx/dcsvg.h:
```diff
@@ -16,7 +16,7 @@
 
```
`#include "wx/scopedptr.h"` seems to be missing here.

MaartenBent replied:
All builds, including those without precompiled headers, succeed. It doesn't seem necessary.

vadz commented on include/wx/dcsvg.h:
```diff
@@ -225,14 +229,14 @@ class WXDLLIMPEXP_CORE wxSVGFileDCImpl : public wxDCImpl
    void DoStartNewGraphics();
 
-   wxFileOutputStream *m_outfile;
+   wxScopedPtr<wxFileOutputStream> m_outfile;
```
Actually, why does this have to be a pointer at all? It looks like it could be just a member as we always have the file name to initialize it with in the ctor...

MaartenBent replied:
Then you have to move the initialization of ```m_outfile``` to the constructor initialization list, instead of using the ```Init()``` function. Next question is, why not put all members in the initialization list and remove the ```Init()``` function completely?

Also, ```"wfstream.h"``` will need to be included in ```dcsvg.h``` and forward declaration becomes obsolete. Is this desired?

vadz commented:
Sorry, I still have a few minor remarks, it would be great if you could please address them. Thanks again!

MaartenBent committed:
Enable usage of clipping regions in wxSVGFileDC.

It was implemented 3 and half years ago in 614e38d.
Close remaining clipping regions when closing the SVG file.

MaartenBent committed:
Correctly draw poly-polygons in wxSVGFileDC.

Each polygon in the poly-polygon needs to end with the start-point. The implementation is similar to the wxDCImpl implementation, except for the extra end points that are inserted.

MaartenBent committed:
Draw lines in wxSVGFileDC as one connected line.

Do not draw each line separately because the lines will not be connected. E.g. when editing a SVG, moving a point should also move the connecting lines.

MaartenBent committed:
Support multi-line text in wxSVGFileDC.

MaartenBent committed:
Improvements to printing text in wxSVGFileDC.

Support underlined and strike-through text.
Set the text length (see http://trac.wxwidgets.org/ticket/17271).
Preserve leading and trailing white-space.
Use wxS macro consistently.
Closes #17271.

MaartenBent committed:
Implemented Clear in wxSVGFileDC.

Draw a rectangle with the background colour and a transparent pen.
Closes #15788.

MaartenBent committed:
Support more pen styles in wxSVGFileDC.

MaartenBent committed:
Support more brush styles in wxSVGFileDC.

MaartenBent committed:
Allow to set the SVG title when creating a wxSVGFileDC.

MaartenBent committed:
Generate valid XML in wxSVGFileDC and updated wxSVGVersion.

Use the correct doc-type and specify the encoding. 'title' is not a valid attribute of <image> so remove it.
Removed superfluous white-space and improved indenting in generated XML.

MaartenBent committed:
Add saving as SVG to drawing sample.

Changed the colour of the third star-polygon to test brush pattern with different colours.
Use the width and height of the scroll panel as image size.

MaartenBent committed:
Improved memory management in wxSVGFileDC.

MaartenBent commented:
Updated. Waiting for input on removing pointer of `m_outfile` (see outdated comment), or if last commit is sufficient.

Relevant switch statements now include all enums. But I wonder if clang will complain about missing `default` case, or about empty statements (and `wxFALLTHROUGH` is needed).

MaartenBent committed:
Support SetDeviceOrigin and SetAxisOrientation in wxSVGFileDC.

Override SetDeviceOrigin, SetLogicalOrigin and SetAxisOrientation from wxDC and mark the graphics as changed, so the correct transform translations are applied.

MaartenBent committed:
Fixed drawing ellipses and arcs in wxSVGFileDC.

Ellipses with the same start and end point (circles) where not drawn because the angle becomes 0 degrees. Fixed by drawing two half circles.
Do not close ellipses if a transparent brush is used (to match wxDC behavior).
See issue #17557.

MaartenBent committed:
Fixed drawing elliptic arcs in wxSVGFileDC.

For some combinations of start and end angles, the wrong large-arc-flag was calculated. Fixed by correctly converting the wxDC angles to SVG angles (shift -90 degrees, and invert to clockwise direction).
Arcs with the same start and end point (circles) where not drawn because the angle becomes 0 degrees. Fixed by drawing two half circles.
Elliptic arcs with a non-transparent brush had an extra line from the center to the start point of the arc. Fixed by first drawing the arc without border, then only the border.
Arcs with small angles would become invisible because the start and end point map to the same (integer) coordinate. Very large arcs would be distorted because the start and end point coordinates did not line up. Using floating point values resolves this.
See issue #17557.

a-wi commented on samples/drawing/drawing.cpp:
```diff
@@ -2169,14 +2175,38 @@ void MyFrame::OnSave(wxCommandEvent& WXUNUSED(event))
+#endif
+            wxSVGFileDC svgdc(dlg.GetPath(), width, height, 72, wxT("Drawing sample"));
+            svgdc.SetBitmapHandler(new wxSVGBitmapEmbedHandler());
+            m_canvas->Draw(svgdc);
```
This trick with swapping DC's doesn't work if "graphics screen" is displayed. In this case an empty image is saved and artefacts are displayed on the screen. It would be good to disable somehow the option to "save as SVG image" if graphics screen is active.

a-wi commented on src/common/dcsvg.cpp:
```diff
@@ -108,6 +119,157 @@ wxString wxBrushString(wxColour c, int style = wxBRUSHSTYLE_SOLID)
+    switch (pen.GetStyle())
+    {
+        case wxPENSTYLE_DOT:
+            s.Printf(wxT("stroke-dasharray=\"%f,%f\" "), w * 2, w * 5);
```
In wxString string literals (at least in these newly introduced) it would be better to use wxS macro.

a-wi commented on src/common/dcsvg.cpp:
```diff
@@ -313,10 +490,27 @@ void wxSVGFileDCImpl::DoDrawLine (wxCoord x1, wxCoord y1, wxCoord x2, wxCoord y2
+            CalcBoundingBox(points[i].x + xoffset, points[i].y + yoffset);
+        }
+
+        s += wxS("\" style=\"fill:none\" /> \n");
```
Is this space between XML tag and EOL really necessary? These trailing spaces are used inconsistently in the code.

MaartenBent committed:
Code cleanup in wxSVGFileDC.

Consistent white-space usage.
Use wxS macros for strings.
Check if output stream is OK when writing.
Removed unnecessary Borland pragmas.

MaartenBent commented:
I updated the commits to use wxS.
Removed all unnecessary white space from generated XML and added some indenting.
Updated the sample to handle graphics screen.

Added an extra commit to clean-up the source code and use wxS instead of wxT.

vadz commented:
If there are no more comments/remarks/objections, I'm going to merge it soon.

Thanks!

wnitzan commented:
The next wxPython release will still pull from WX_3_0_BRANCH.
Will it be possible for you to back-port these SVG changes into WX_3_0_BRANCH?

MaartenBent commented:
Sure, I can create a PR for 3.0 branch with all recent SVG changes.

After comparing the headers I do have some concerns about ABI/API compatibility. I'll address them in the new PR.

wnitzan commented:
I can't thank you enough. These SVG fixes are badly needed on the wxPython side.
Please come back to update here with your efforts, so I can follow your progress.
